packages/itwin/tree-widget/src/components/trees/stateless/models-tree/HierarchyVisibilityHandler.ts

```diff
@@ -0,0 +1,877 @@
+  }
+
+  private getElementGroupingNodeDisplayStatus(node: GroupingHierarchyNode): Observable<VisibilityStatus> {
+    const result = this.getGroupedElementIds(node).pipe(
+      mergeMap((group) => {
+        if (!group) {
+          return createVisibilityStatusObs("disabled");
+        }
+
+        const { modelId, categoryId, elementIds } = group;
+        if (!this._props.viewport.view.viewsModel(modelId)) {
+          return createVisibilityStatusObs("hidden");
+        }
+
+        return elementIds.pipe(
+          toSet(),
+          mergeMap((ids) => {
+            return this.getVisibilityFromAlwaysAndNeverDrawnElements({
+              elements: ids,
+              defaultStatus: () => {
+                const status = this.getDefaultCategoryVisibilityStatus(categoryId, modelId);
+                return createVisibilityStatus(status.state, `groupingNode.${status.state}ThroughCategory`);
+              },
+              tooltips: {
+                allElementsInAlwaysDrawnList: "groupingNode.allElementsVisible",
+                allElementsInNeverDrawnList: "groupingNode.allElementsHidden",
+                elementsInBothAlwaysAndNeverDrawn: "groupingNode.someElementsAreHidden",
+                noElementsInExclusiveAlwaysDrawnList: "groupingNode.allElementsHidden",
+              },
+            });
+          }),
+        );
+      }),
+    );
+
+    const ovr = this._props.overrides?.getElementGroupingNodeDisplayStatus;
+    return ovr ? from(ovr(this.createOverrideProps({ node }, result))) : result;
+  }
+
+  private getElementOverriddenVisibility(elementId: string): NonPartialVisibilityStatus | undefined {
+    const viewport = this._props.viewport;
+    if (viewport.neverDrawn?.has(elementId)) {
+      return createVisibilityStatus("hidden", "element.hiddenThroughNeverDrawnList");
+    }
+
+    if (viewport.alwaysDrawn?.size) {
+      if (viewport.alwaysDrawn.has(elementId)) {
+        return createVisibilityStatus("visible", "element.displayedThroughAlwaysDrawnList");
+      }
+
+      if (viewport.isAlwaysDrawnExclusive) {
+        return createVisibilityStatus("hidden", "element.hiddenDueToOtherElementsExclusivelyAlwaysDrawn");
+      }
+    }
+
+    return undefined;
+  }
+
+  private getElementDisplayStatus(props: GetElementStateProps): Observable<VisibilityStatus> {
+    const result = defer(() => {
+      const viewport = this._props.viewport;
+      const { elementId, modelId, categoryId } = props;
+
+      if (!viewport.view.viewsModel(modelId)) {
+        return createVisibilityStatusObs("hidden", "element.hiddenThroughModel");
+      }
+
+      let status = this.getElementOverriddenVisibility(elementId);
+      if (status) {
+        return of(status);
+      }
+
+      status = this.getDefaultCategoryVisibilityStatus(categoryId, modelId);
+      return createVisibilityStatusObs(status.state, status.state === "visible" ? "element.visibleThroughCategory" : "hiddenThroughCategory");
```

Comment:
Fixed (P.S i wonder if there's a way to import this JSON as a TS type so that these mistakes could be avoided)

Reply:
> i wonder if there's a way to import this JSON as a TS type so that these mistakes could be avoided

I don't think so, since the JSON is in format 
```json
{
  "element": {
    "hiddenThroughCategory": "value"
  }
}
```
and we're using it in format `"element.hiddenThroughCategory"`. Maybe it's possible to do some type mapping, but I guess it's just easier to avoid these mistakes through e2e tests.